### DISCUSSION TITLE
Add max amount to fields: tags, checkboxes, table

### DISCUSSION THREAD

**Discussion by angrybrad** [craftcms/cms#9169] [Ideas]
> Created by: Rob (rob@layermade.com) on 2015/07/19 22:22:24 +0000
> Votes at time of UserVoice import: 6

-----------------------------------

I've come up with this issue a couple of times, where from a UX point of view we need to limit how many options a user can select: tags, checkbox, table rows. I have the specific use-case where the client wants selectable tags, but limiting them within a user group.

Ideally you'd be able to do this depending on it's context (admin, user-group, section), but I guess that's going to be difficult to achieve.

The only option I can think of at the moment is to use something like Chosen on a front-end template: http://bit.ly/1VuMdy9

**Comment by brandonkelly** [comment]
This is still a valid feature request, but in the meantime, you can add your own custom validation rules to entries using the [EVENT_DEFINE_RULES](https://docs.craftcms.com/api/v3/craft-base-model.html#event-define-rules) event from a custom [module](https://docs.craftcms.com/v3/extend/module-guide.html).

And Craft provides an [ArrayValidator](https://docs.craftcms.com/api/v3/craft-validators-arrayvalidator.html) validation class, which makes it easy to validate based on the size of an array (or element query result, as in the case of a Tags field).

```php
use craft\elements\Entry;
use craft\events\DefineRulesEvent;
use craft\validators\ArrayValidator;
use yii\base\Event;

Event::on(Entry::class, Entry::EVENT_DEFINE_RULES, function(DefineRulesEvent $e) {
    /** @var Entry $entry */
    $entry = $e->sender;

    // Only worry about entries in the News section
    if ($entry->section->handle !== 'news') {
        return;
    }

    $e->rules[] = ['field:myTagsField', ArrayValidator::class, 'max' => 5, 'on' => Entry::SCENARIO_LIVE];
    $e->rules[] = ['field:myCheckboxesField', ArrayValidator::class, 'max' => 5, 'on' => Entry::SCENARIO_LIVE];
    $e->rules[] = ['field:myTableField', ArrayValidator::class, 'max' => 5, 'on' => Entry::SCENARIO_LIVE];
});
```

**Reply by proimage** [reply]
@brandonkelly This wouldn't hide the "Add a tag" prompt after adding 5 tags, would it?

**Reply by proimage** [reply]
This is the use-case, BTW... a Super-Table field with Tag field columns, to allow the client to customize the link tracking parameters on social media links:

![image](https://user-images.githubusercontent.com/1027523/165230691-0518b478-902d-4a1c-b6f3-8e2586be5617.png)

The Tag field experience is preferred for this because it both allows them to freely customize both parameter and value, yet helps prevent typos by prompting them with existing items that match what they're typing. The only problem with this setup is that you can add more than one tag to a tag field:

![image](https://user-images.githubusercontent.com/1027523/165230995-f1f6d532-af69-4558-855d-3f057d07ce8e.png)

Alternately, if Craft had a [datalist field](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/datalist), that would work perfectly as well (for my use-case, at least).
